File: regions/system/library/dart/pagination/test/pagination_test.dart
import 'package:test/test.dart';

import 'package:k1s0_pagination/pagination.dart';

void main() {
  group('PageRequest', () {
    test('pageとperPageが保持されること', () {
      const req = PageRequest(page: 1, perPage: 10);
      expect(req.page, equals(1));
      expect(req.perPage, equals(10));
    });

    test('defaultRequestがpage=1, perPage=20を返すこと', () {
      final req = PageRequest.defaultRequest();
      expect(req.page, equals(1));
      expect(req.perPage, equals(20));
    });

    test('offsetが(page-1) * perPageで計算されること', () {
      const req1 = PageRequest(page: 1, perPage: 10);
      expect(req1.offset, equals(0));

      const req2 = PageRequest(page: 3, perPage: 10);
      expect(req2.offset, equals(20));

      const req3 = PageRequest(page: 2, perPage: 25);
      expect(req3.offset, equals(25));
    });

    test('次のページが存在する場合hasNextがtrueを返すこと', () {
      const req = PageRequest(page: 1, perPage: 10);
      expect(req.hasNext(25), isTrue);
      expect(req.hasNext(10), isFalse);
      expect(req.hasNext(11), isTrue);
    });

    test('最終ページではhasNextがfalseを返すこと', () {
      const req = PageRequest(page: 3, perPage: 10);
      expect(req.hasNext(30), isFalse);
      expect(req.hasNext(31), isTrue);
    });
  });

  group('PageResponse', () {
    test('正しいtotalPagesで生成されること', () {
      const req = PageRequest(page: 1, perPage: 10);
      final resp = PageResponse<String>.create(['a', 'b'], 25, req);
      expect(resp.items, equals(['a', 'b']));
      expect(resp.total, equals(25));
      expect(resp.totalPages, equals(3));
    });

    test('割り切れる件数の場合に正しいページ数が計算されること', () {
      const req = PageRequest(page: 1, perPage: 5);
      final resp = PageResponse<int>.create([1, 2, 3, 4, 5], 10, req);
      expect(resp.totalPages, equals(2));
    });

    test('アイテムが空の場合に正しく処理されること', () {
      const req = PageRequest(page: 1, perPage: 10);
      final resp = PageResponse<String>.create([], 0, req);
      expect(resp.totalPages, equals(0));
      expect(resp.items, isEmpty);
    });

    test('metaがPaginationMetaを返すこと', () {
      const req = PageRequest(page: 2, perPage: 10);
      final resp = PageResponse<String>.create(['a'], 25, req);
      final meta = resp.meta;
      expect(meta.total, equals(25));
      expect(meta.page, equals(2));
      expect(meta.perPage, equals(10));
      expect(meta.totalPages, equals(3));
    });
  });

  group('cursor', () {
    test('エンコードとデコードのラウンドトリップが正しく動作すること', () {
      const sortKey = '2024-01-15';
      const id = 'abc-123';
      final cursor = encodeCursor(sortKey, id);
      final result = decodeCursor(cursor);
      expect(result.sortKey, equals(sortKey));
      expect(result.id, equals(id));
    });

    test('base64url形式の文字列が生成されること', () {
      final cursor = encodeCursor('key', 'test-id');
      expect(cursor, isNotEmpty);
      expect(cursor, isNot(contains(' ')));
    });

    test('CursorRequestがフィールドを保持すること', () {
      const req = CursorRequest(cursor: 'abc', limit: 20);
      expect(req.cursor, equals('abc'));
      expect(req.limit, equals(20));
    });

    test('CursorMetaがフィールドを保持すること', () {
      const meta = CursorMeta(nextCursor: 'next', hasMore: true);
      expect(meta.nextCursor, equals('next'));
      expect(meta.hasMore, isTrue);
    });
  });

  group('validatePerPage', () {
    test('有効な値を受け付けること', () {
      expect(validatePerPage(1), equals(1));
      expect(validatePerPage(50), equals(50));
      expect(validatePerPage(100), equals(100));
    });

    test('0を拒否すること', () {
      expect(() => validatePerPage(0), throwsA(isA<PerPageValidationException>()));
    });

    test('最大値を超えた場合に拒否すること', () {
      expect(() => validatePerPage(101), throwsA(isA<PerPageValidationException>()));
    });
  });
}

File: regions/system/library/go/pagination/pagination_test.go
package pagination_test

import (
	"testing"

	"github.com/k1s0-platform/system-library-go-pagination"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// NewPageResponseが総件数とページサイズから総ページ数を正しく計算することを確認する。
func TestNewPageResponse_TotalPages(t *testing.T) {
	items := []string{"a", "b", "c"}
	resp := pagination.NewPageResponse(items, 10, pagination.PageRequest{Page: 1, PerPage: 3})
	assert.Equal(t, uint32(4), resp.TotalPages)
	assert.Equal(t, uint32(1), resp.Page)
	assert.Equal(t, uint32(3), resp.PerPage)
	assert.Equal(t, uint64(10), resp.Total)
	assert.Equal(t, 3, len(resp.Items))
}

// 総件数がページサイズで割り切れる場合にTotalPagesが正確に計算されることを確認する。
func TestNewPageResponse_ExactDivision(t *testing.T) {
	items := []int{1, 2}
	resp := pagination.NewPageResponse(items, 6, pagination.PageRequest{Page: 3, PerPage: 2})
	assert.Equal(t, uint32(3), resp.TotalPages)
}

// PerPageに0を指定した場合にデフォルト値1が使われ正しくTotalPagesが計算されることを確認する。
func TestNewPageResponse_ZeroPerPage(t *testing.T) {
	resp := pagination.NewPageResponse([]int{}, 5, pagination.PageRequest{Page: 1, PerPage: 0})
	assert.Equal(t, uint32(5), resp.TotalPages)
	assert.Equal(t, uint32(1), resp.PerPage)
}

// PageResponseのMetaメソッドがページネーションメタ情報を正しく返すことを確認する。
func TestPageResponse_Meta(t *testing.T) {
	resp := pagination.NewPageResponse([]int{1, 2, 3}, 25, pagination.PageRequest{Page: 2, PerPage: 10})
	meta := resp.Meta()
	assert.Equal(t, uint64(25), meta.Total)
	assert.Equal(t, uint32(2), meta.Page)
	assert.Equal(t, uint32(10), meta.PerPage)
	assert.Equal(t, uint32(3), meta.TotalPages)
}

// EncodeCursorとDecodeCursorがカーソルのエンコードとデコードを正しく往復できることを確認する。
func TestCursor_RoundTrip(t *testing.T) {
	sortKey := "2024-01-15"
	id := "abc-123-def"
	encoded := pagination.EncodeCursor(sortKey, id)
	decodedSortKey, decodedID, err := pagination.DecodeCursor(encoded)
	require.NoError(t, err)
	assert.Equal(t, sortKey, decodedSortKey)
	assert.Equal(t, id, decodedID)
}

// 不正なbase64文字列に対してDecodeCursorがエラーを返すことを確認する。
func TestDecodeCursor_Invalid(t *testing.T) {
	_, _, err := pagination.DecodeCursor("!!!invalid!!!")
	assert.Error(t, err)
}

// パイプ区切り文字を含まないカーソルに対してDecodeCursorがエラーを返すことを確認する。
func TestDecodeCursor_MissingSeparator(t *testing.T) {
	// base64("noseparator") - contains no pipe separator
	_, _, err := pagination.DecodeCursor("bm9zZXBhcmF0b3I=")
	assert.Error(t, err)
}

// ValidatePerPageが有効な範囲の値（1〜100）に対してエラーなしを返すことを確認する。
func TestValidatePerPage_Valid(t *testing.T) {
	assert.NoError(t, pagination.ValidatePerPage(1))
	assert.NoError(t, pagination.ValidatePerPage(50))
	assert.NoError(t, pagination.ValidatePerPage(100))
}

// ValidatePerPageが0を指定した場合にエラーを返すことを確認する。
func TestValidatePerPage_Zero(t *testing.T) {
	assert.Error(t, pagination.ValidatePerPage(0))
}

// ValidatePerPageが最大値（100）を超えた場合にエラーを返すことを確認する。
func TestValidatePerPage_OverMax(t *testing.T) {
	assert.Error(t, pagination.ValidatePerPage(101))
}

// CursorRequestのCursorとLimitフィールドが正しく設定されることを確認する。
func TestCursorRequest_Fields(t *testing.T) {
	cursor := "some-cursor"
	req := pagination.CursorRequest{Cursor: &cursor, Limit: 20}
	assert.Equal(t, &cursor, req.Cursor)
	assert.Equal(t, uint32(20), req.Limit)
}

// CursorMetaのNextCursorとHasMoreフィールドが正しく設定されることを確認する。
func TestCursorMeta_Fields(t *testing.T) {
	next := "next-cursor"
	meta := pagination.CursorMeta{NextCursor: &next, HasMore: true}
	assert.Equal(t, &next, meta.NextCursor)
	assert.True(t, meta.HasMore)
}

// PaginationMetaの各フィールドが正しく設定されることを確認する。
func TestPaginationMeta_Fields(t *testing.T) {
	meta := pagination.PaginationMeta{
		Total:      100,
		Page:       2,
		PerPage:    10,
		TotalPages: 10,
	}
	assert.Equal(t, uint64(100), meta.Total)
	assert.Equal(t, uint32(10), meta.TotalPages)
}

// NewPageRequestが指定したページ番号とページサイズを持つPageRequestを生成することを確認する。
func TestNewPageRequest(t *testing.T) {
	req := pagination.NewPageRequest(3, 50)
	assert.Equal(t, uint32(3), req.Page)
	assert.Equal(t, uint32(50), req.PerPage)
}

// DefaultPageRequestがデフォルトのPage=1、PerPage=20を返すことを確認する。
func TestDefaultPageRequest(t *testing.T) {
	req := pagination.DefaultPageRequest()
	assert.Equal(t, uint32(1), req.Page)
	assert.Equal(t, uint32(20), req.PerPage)
}

// PageRequestのOffsetがページ番号とページサイズから正しいオフセット値を返すことを確認する。
func TestPageRequest_Offset(t *testing.T) {
	assert.Equal(t, uint64(0), pagination.NewPageRequest(1, 20).Offset())
	assert.Equal(t, uint64(20), pagination.NewPageRequest(2, 20).Offset())
	assert.Equal(t, uint64(40), pagination.NewPageRequest(3, 20).Offset())
	assert.Equal(t, uint64(0), pagination.DefaultPageRequest().Offset())
}

// PageRequestのHasNextが総件数に基づいて次ページの有無を正しく判定することを確認する。
func TestPageRequest_HasNext(t *testing.T) {
	req := pagination.NewPageRequest(1, 10)
	assert.True(t, req.HasNext(11))
	assert.False(t, req.HasNext(10))
	assert.False(t, req.HasNext(5))

	req2 := pagination.NewPageRequest(2, 10)
	assert.True(t, req2.HasNext(21))
	assert.False(t, req2.HasNext(20))
}
